node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,379 @@
+                                // a network message was received
+                                if let Err(e) = handle_network_msg(&mut ctx, &mut tracker, event).await {
+                                    log::warn!("Failed to handle incomming network messages: {:?}", e);
+                                }
+                            }
+                        }
+                    }
+                    FromOverseer::Signal(OverseerSignal::StartWork(relay_parent)) => {
+                        // query basic system parameters once
+                        // @todo assumption: these cannot change within a session
+                        let (validator_set, signing_context) =
+                            query_basics(&mut ctx, relay_parent).await?;
+
+                        let _ = tracker.per_relay_parent.insert(
+                            relay_parent,
+                            PerRelayParentData {
+                                signing_context,
+                                validator_set,
+                                ..Default::default()
+                            },
+                        );
+                    }
+                    FromOverseer::Signal(OverseerSignal::StopWork(relay_parent)) => {
+                        // @todo assumption: it is good enough to prevent addition work from being
+                        // scheduled, the individual futures are supposedly completed quickly
+                        let _ = tracker.per_relay_parent.remove(&relay_parent);
+                    }
+                    FromOverseer::Signal(OverseerSignal::Conclude) => {
+                        tracker.per_relay_parent.clear();
+                        return Ok(());
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Modify the reputiation of peer based on their behaviour.
+async fn modify_reputiation<Context>(
+    ctx: &mut Context,
+    peerid: PeerId,
+    rep: ReputationChange,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+    ctx.send_message(AllMessages::NetworkBridge(
+        NetworkBridgeMessage::ReportPeer(peerid, rep),
+    ))
+    .await
+}
+
+/// Distribute a given valid bitfield message.
+///
+/// Can be originated by another subsystem or received via network from another peer.
+async fn distribute<Context>(
+    ctx: &mut Context,
+    tracker: &mut Tracker,
+    message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+    let BitfieldGossipMessage {
+        relay_parent,
+        signed_availability,
+    } = message;
+    // concurrently pass on the bitfield distribution to all interested peers
+    let interested_peers = tracker
+        .peer_views
+        .iter()
+        .filter_map(|(peerid, view)| {
+            if view.contains(&relay_parent) {
+                Some(peerid.clone())
```

Comment:
should also have a check that the peer hasn't yet seen anything by that validator, right?

Reply:
(a malicious validator could issue 2+ bitfields and introduce them at separate points of the network. if we lie on that boundary, the peer may send us 1 and we would send them the other). I think we need to track if the peer has already sent us a bitfield per validator so we can punish them if they ever send another. And then we should track if we've sent a bitfield per validator.

We don't do equivocation reporting or anything like that for bitfields, though.